server/applicationset/applicationset.go

```diff
@@ -0,0 +1,366 @@
+		if err := s.enf.EnforceErr(ctx.Value("claims"), rbacpolicy.ResourceApplicationSets, rbacpolicy.ActionUpdate, apputil.AppSetRBACName(appset)); err != nil {
+			return nil, err
+		}
+		// They also need 'delete' privileges in the old project
+		if err := s.enf.EnforceErr(ctx.Value("claims"), rbacpolicy.ResourceApplicationSets, rbacpolicy.ActionDelete, apputil.AppSetRBACName(appset)); err != nil {
+			return nil, err
+		}
+	}
+
+	for i := 0; i < 10; i++ {
+		appset.Spec = newAppset.Spec
+		if merge {
+			appset.Labels = mergeStringMaps(appset.Labels, newAppset.Labels)
+			appset.Annotations = mergeStringMaps(appset.Annotations, newAppset.Annotations)
+		} else {
+			appset.Labels = newAppset.Labels
+			appset.Annotations = newAppset.Annotations
+		}
+
+		res, err := s.appclientset.ArgoprojV1alpha1().ApplicationSets(s.ns).Update(ctx, appset, metav1.UpdateOptions{})
+		if err == nil {
+			s.logAppSetEvent(appset, ctx, argo.EventReasonResourceUpdated, "updated ApplicationSets spec")
+			s.waitSync(res)
+			return res, nil
+		}
+		if !apierr.IsConflict(err) {
+			return nil, err
+		}
+
+		appset, err = s.appclientset.ArgoprojV1alpha1().ApplicationSets(s.ns).Get(ctx, newAppset.Name, metav1.GetOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("error getting ApplicationSets: %w", err)
+		}
+	}
+	return nil, status.Errorf(codes.Internal, "Failed to update ApplicationSets. Too many conflicts")
+}
+
+func (s *Server) Delete(ctx context.Context, q *applicationset.ApplicationSetDeleteRequest) (*applicationset.ApplicationSetResponse, error) {
+
+	appset, err := s.appclientset.ArgoprojV1alpha1().ApplicationSets(s.ns).Get(ctx, q.Name, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("error getting ApplicationSets: %w", err)
+	}
+
+	s.projectLock.RLock(appset.Spec.Template.Spec.Project)
+	defer s.projectLock.RUnlock(appset.Spec.Template.Spec.Project)
+
+	if err := s.enf.EnforceErr(ctx.Value("claims"), rbacpolicy.ResourceApplicationSets, rbacpolicy.ActionDelete, apputil.AppSetRBACName(appset)); err != nil {
+		return nil, err
+	}
+
+	err = s.appclientset.ArgoprojV1alpha1().ApplicationSets(s.ns).Delete(ctx, q.Name, metav1.DeleteOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("error deleting ApplicationSets: %w", err)
+	}
+	s.logAppSetEvent(appset, ctx, argo.EventReasonResourceDeleted, "deleted ApplicationSets")
+	return &applicationset.ApplicationSetResponse{}, nil
+
+}
+
+func (s *Server) validateAppSet(ctx context.Context, appset *v1alpha1.ApplicationSet) (string, error) {
+	if appset == nil {
+		return "", fmt.Errorf("ApplicationSet cannot be validated for nil value")
+	}
+
+	projectName := appset.Spec.Template.Spec.Project
+
+	if strings.Contains(projectName, "{{") {
+		return "", fmt.Errorf("the Argo CD API does not currently support creating ApplicationSets with templated `project` fields")
+	}
+
+	if err := appsetutils.CheckInvalidGenerators(appset); err != nil {
+		return "", err
```

Comment:
Wrap this error?

Reply:
The error is already wrapped within the `CheckInvalidGenerators` function, thus not rewrapping it.